api/filters/replacement/replacement.go

```diff
@@ -0,0 +1,114 @@
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+type Filter struct {
+	Replacements []types.Replacement
+}
+
+const DefaultFieldRef = ".metadata.name"
+
+// Filter replaces values of targets with values from sources
+// TODO (#3492): Connect this to a replacement transformer plugin
+func (f Filter) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	for _, r := range f.Replacements {
+		value, err := getReplacement(nodes, r.Source)
+		if err != nil {
+			return nil, err
+		}
+		nodes, err = applyReplacement(nodes, value, r.Target)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nodes, nil
+}
+
+func applyReplacement(nodes []*yaml.RNode, value *yaml.RNode, target *types.ReplTarget) ([]*yaml.RNode, error) {
+	var result []*yaml.RNode
+	for _, n := range nodes {
+		node, err := applyToNode(n, value, target)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, node)
+	}
+	return result, nil
+}
+
+func applyToNode(node *yaml.RNode, value *yaml.RNode, target *types.ReplTarget) (*yaml.RNode, error) {
+	for i := range target.FieldRefs {
+		fieldref := strings.Split(target.FieldRefs[i], ".")
+		// TODO (#3492): Support using map keys in the fieldref (e.g. .spec.containers[name=nginx])
+		t, err := node.Pipe(yaml.Lookup(fieldref...))
+		if err != nil {
+			return nil, err
+		}
+		if t != nil {
+			t.SetYNode(value.YNode())
+		}
+	}
+	return node, nil
+}
+
+func getReplacement(nodes []*yaml.RNode, source *types.ReplSource) (*yaml.RNode, error) {
+	if source.Value != "" {
+		if source.ObjRef != nil || source.FieldRef != "" {
+			return nil, fmt.Errorf("objref and fieldref must be empty if source value is provided")
+		}
+		return yaml.Parse(source.Value)
+	}
+
+	match, err := getMatch(nodes, source.ObjRef)
+	if err != nil {
+		return nil, err
+	}
+
+	fieldRef := source.FieldRef
+	if fieldRef == "" {
+		fieldRef = DefaultFieldRef
+	}
+	fieldRefSlice := strings.Split(fieldRef, ".")
+
+	// TODO (#3492): Support using map keys in the fieldref (e.g. .spec.containers[name=nginx])
+	rn, err := match.Pipe(yaml.Lookup(fieldRefSlice...))
+	if err != nil {
+		return nil, err
+	}
+	return rn, nil
+}
+
+// getMatch finds the node that matches the objRef, throwing an error if none are found or if
+// more than one match is found
+func getMatch(nodes []*yaml.RNode, objRef *types.Target) (*yaml.RNode, error) {
+	var matches []*yaml.RNode
+	for _, n := range nodes {
+		ns, err := n.GetNamespace()
+		if err != nil {
+			return nil, err
+		}
+		apiVersion := yaml.GetValue(n.Field(yaml.APIVersionField).Value)
+		if (n.GetName() == objRef.Name || objRef.Name == "") &&
+			(n.GetKind() == objRef.Kind || objRef.Kind == "") &&
+			(ns == objRef.Namespace || objRef.Namespace == "") &&
+			(apiVersion == objRef.ApiVersion() || objRef.ApiVersion() == "") {
+			matches = append(matches, n)
+		}
+	}
+	if len(matches) > 1 {
+		return nil, fmt.Errorf("found more than one resources matching from %v", nodes)
```

Comment:
Did you mean to print `obRef` here rather than `nodes`, like below? I'm guessing the error message with nodes would be pretty overwhelming

Reply:
This makes me realize there's no error case coverage yet. Can we add some? I'd say this and the source field conflict are the important ones. (Follow-up is fine.)

Reply:
I've changed it to print out objref rather than nodes, thanks for the suggestion. 

I agree there should be error case coverage, but I would like to do that in a follow-up PR. The RunFilter function (used by all filters for their tests) currently panics if there is an error.  Modifying it to be able to accept cases with errors will be a decently sized change that I would like to do separately.

Btw, thank you for your quick reviews!

Reply:
I've updated it to print objRef rather than nodes, and added a couple of error test cases. Thanks for the suggestions!